### PR TITLE
Don't call `getenv` in side threads

### PR DESCRIPTION
**Pull request by eqy**
Calling `getenv` on side threads is dangerous as it can potentially segfault if the main thread is in the middle of setting environment variables: https://github.com/pytorch/pytorch/issues/134596

This PR only calls `getenv` only once during the first call of  `isDaemonEnvVarSet()`, which is called from `init`.

Co-authored-by: Nikita Shulga <2453524+malfet@users.noreply.github.com>

### CODE REVIEW

**Review by nWEIdia — APPROVED** [review]
Genius PR!

**Review by aaronenyeshi — APPROVED** [review]

**Review by malfet — APPROVED** [review]

File: libkineto/include/Config.h [hunk 1]
```diff
@@ -502,4 +502,8 @@ class Config : public AbstractConfig {
 
 constexpr char kUseDaemonEnvVar[] = "KINETO_USE_DAEMON";
 
+#if __linux__
+extern int kUseDaemonEnvVarSet;
```
Comment (malfet) [inline libkineto/include/Config.h:506]:
```suggestion
extern bool isDaemonEnvVarSet();
```

File: libkineto/src/Config.cpp [hunk 2]
```diff
@@ -243,7 +244,8 @@ Config::Config()
     factories->addFeatureConfigs(*this);
   }
 #if __linux__
-  enableIpcFabric_ = (getenv(kUseDaemonEnvVar) != nullptr);
+  assert(libkineto::kUseDaemonEnvVarSet >= 0);
+  enableIpcFabric_ = libkineto::kUseDaemonEnvVarSet;
```
Comment (malfet) [inline libkineto/src/Config.cpp:248]:
```suggestion
  enableIpcFabric_ = libkineto::isDaemonEnvVarSet();
```

File: libkineto/src/init.cpp [hunk 3]
```diff
@@ -31,6 +31,10 @@
 #include "Logger.h"
 
 namespace KINETO_NAMESPACE {
+#if __linux__
+int kUseDaemonEnvVarSet = -1;
```
Comment (malfet) [inline libkineto/src/init.cpp:35]:
```suggestion
bool isDaemonEnvVarSet() {
  static bool rc = [] {
      void *ptr = getenv(kUseDaemonEnvVar);
      return ptr != nullptr;
  }();
  return rc;
}
```

File: libkineto/src/init.cpp [hunk 4]
```diff
@@ -131,7 +135,9 @@ void libkineto_init(bool cpuOnly, bool logOnError) {
 
   // Factory to connect to open source daemon if present
 #if __linux__
-  if (getenv(kUseDaemonEnvVar) != nullptr) {
+  void *ptr = getenv(kUseDaemonEnvVar);
+  libkineto::kUseDaemonEnvVarSet = ptr != nullptr;
+  if (libkineto::kUseDaemonEnvVarSet) {
```
Comment (malfet) [inline libkineto/src/init.cpp:140]:
```suggestion
  if (libkineto::isDaemonEnvVarSet()) {
```

File: libkineto/include/Config.h [hunk 5]
```diff
@@ -502,4 +502,8 @@ class Config : public AbstractConfig {
 
 constexpr char kUseDaemonEnvVar[] = "KINETO_USE_DAEMON";
 
+#if __linux__
+extern bool isDaemonEnvVarSet();
```
Comment (malfet) [inline libkineto/include/Config.h:506]:
```suggestion
bool isDaemonEnvVarSet();
```

**Review by sraikund16 — APPROVED** [review]
LGTM!

File: libkineto/src/Config.cpp [hunk 6]
```diff
@@ -8,6 +8,7 @@
 
 #include "Config.h"
 
+#include <assert.h>
```
Comment (malfet) [inline libkineto/src/Config.cpp:11]:
```suggestion
```

File: libkineto/src/init.cpp [hunk 7]
```diff
@@ -31,6 +31,16 @@
...
+#if __linux__
+bool isDaemonEnvVarSet() {
+  static bool rc = [] {
+      void *ptr = getenv(kUseDaemonEnvVar);
+      return ptr != nullptr;
+  }();
+  return rc;
+}
+#endif
+
```
Comment (malfet) [inline libkineto/src/init.cpp:43]:
```suggestion
```

File: libkineto/src/Config.cpp [hunk 8]
```diff
@@ -243,7 +243,7 @@ Config::Config()
     factories->addFeatureConfigs(*this);
   }
 #if __linux__
-  enableIpcFabric_ = (getenv(kUseDaemonEnvVar) != nullptr);
+  enableIpcFabric_ = libkineto::isDaemonEnvVarSet();
 #endif
 }
```
Comment (malfet) [inline libkineto/src/Config.cpp:248]:
```suggestion
}

#if __linux__
bool isDaemonEnvVarSet() {
  static bool rc = [] {
      void *ptr = getenv(kUseDaemonEnvVar);
      return ptr != nullptr;
  }();
  return rc;
}
#endif

```

**Pull request closed without merging**